### PR TITLE
Rtd buildout

### PR DESCRIPTION
**Pull request by Phrozyn**
This is ready for review

### CODE REVIEW

File: docs/source/overview.rst
Comment (caggle) [inline docs/source/overview.rst]: Use REST instead of "rest".

File: docs/source/overview.rst
Comment (caggle) [inline docs/source/overview.rst]: There is nothing factually wrong with this diagram, but I thought in the past we discussed having separate flow diagrams for each option: CB only, Github only, both with merge.

File: docs/source/overview.rst
Comment (caggle) [inline docs/source/overview.rst]: Sorry to nitpick, but I think the correct term is SSM Parameter Store.

File: docs/source/overview.rst
Comment (caggle) [inline docs/source/overview.rst]: There is a typo here (rEST).

File: docs/source/architecture.rst
Comment (caggle) [inline docs/source/architecture.rst]: Again nitpicking: We should use S3 instead of s3.

File: docs/source/architecture.rst
Comment (caggle) [inline docs/source/architecture.rst]: Since we now have multiple options, we should probably say "...while the only difference will be the source code repository" instead of CodeCommit.

File: docs/source/architecture.rst
Comment (caggle) [inline docs/source/architecture.rst]: Similar to my above comment...To keep it generic we should probably say "...hat are populated by the buildspec.yml located in each event source configuration."

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst]: the / represents or, meaning either can be used here.

File: docs/source/architecture.rst
Comment (Phrozyn) [reply docs/source/architecture.rst]: fixed! Ty!

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst]: Fixed! TY!

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst]: Fixed! TY!

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst]: We will still have them, but I haven't created them yet. This is still a WIP.

File: docs/source/architecture.rst
Comment (Phrozyn) [reply docs/source/architecture.rst]: I'll resolve this with the use of footnotes to inform.

File: docs/source/architecture.rst
Comment (Phrozyn) [reply docs/source/architecture.rst]: This has been revised, please review.

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst]: Per our conversation I'll fix.

Comment (Phrozyn) [reply docs/source/overview.rst]: resolved by changing to SSM and Secrets manager parameters.

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst]: Working on these diagrams now

Comment (Phrozyn) [reply docs/source/overview.rst]: We don't need separate diagrams for each, because the workflow for using a single repo is the same, only the repo source is different. I believe this creates redundancy. 
So we'll only have two diagrams, one for multi-source and one for single source.

We'll just need to advise in the directions any differences.

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: API_PATH should not be `/`, but some other string. If `/` is specified here, the URL for the API would end in double slash `//`, which is not what we want.

Maybe we can change it to `zoom` or simply `api`.

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: I wonder, should we generalise and say "For webhook-based event sources (such as Zoom), there should not be any other changes." instead of "For the zoom service, ...".

Just an idea.

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: We should change the end of this sentence to `"... MozDef-EF-<environment>."`

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: Maybe here say: `<service>2mozdef`.

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: Here we can say: `For service, enter the value decided for the service in "Getting Started" Step 2.

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: We should probably clarify this a bit more.

The default PLBucketPrefix value is `mozdef-ef-utility-bucket`. As discussed, we can remove the "bucket" from the name. Assume we ditch it, I think we should change this line to something like:

"Enter a name for the S3 bucket to be used by AWS CodePipeline to store artifacts. You can leave the default value as is, and this would create a bucket for you with the name `mozdef-ef-utility-<stackname>-<environment>`."

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: Maybe use "monitor" instead of "watch".

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: yeah I didn't know what this should be, so just added a / for now.
I'll craft an example path

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: Haha you wrote that, I just copied and pasted :p

We can definitely modify, this is exactly the feedback I was looking for  :) ty

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: or `"...<project>-<environment>."`

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: yes, ok!

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: well mozdef-ef is actually the `<project>` name, so it would be more like 

`<project>-utility-<stackname>-<environment>`

However I feel like that's really long, and should possibly be tied to the service using it rather than the stackname, so 

`<project>-utility-<service>-<environment>`

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: done

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: check out what I've added and let me know if that works for you :)

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: review changes :)

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: hmm maybe that's not possible, cause it should probably have some way to tie back to the stack name.

Comment (caggle) [reply docs/source/usage/deployment.rst]: I think using stack name and environment is OK.

Also we should note here that whether or not we should use <project> as the suffix is only for documentation purposes, as we are not specifying <project> as a stack parameter (this is why it was hardcoded as `mozdef-ef...` in the CF templates. <project> value is only specified in buildspec as an environment variable, and used by the serverless framework/app, not the pipeline stack.

If we want to avoid this chicken and egg kind of problem, we can just make the default name of the bucket in the CF template something like `<stackname>-<environment>-utility`. I think this would be descriptive enough.

File: docs/source/overview.rst
Comment (caggle) [reply docs/source/overview.rst]: This works for me. Thanks.

File: docs/source/architecture.rst
Comment (caggle) [reply docs/source/architecture.rst]: Yeah, sounds good. Ty!

File: docs/source/usage/deployment.rst
Comment (caggle) [reply docs/source/usage/deployment.rst]: Yeah I know I wrote the most of this originally :) This is a way for me to review what I wrote as well. I figured, before it was documented rather less formally on a Gdoc, and a proper documentation deserves a better explanation.

File: docs/source/usage/deployment.rst
Comment (caggle) [reply docs/source/usage/deployment.rst]: Looks good.

File: docs/source/usage/deployment.rst
Comment (caggle) [reply docs/source/usage/deployment.rst]: Yeah changes look great. Ty!

File: docs/source/usage/deployment.rst
Comment (caggle) [reply docs/source/usage/deployment.rst]: I pushed this change to `master`. Please let me know if this was a mistake or not something we agreed on.

Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: This worked perfectly! ty

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: Fixed

File: docs/source/overview.rst
Comment (caggle) [inline docs/source/overview.rst]: Typo here (there vs. three)

File: docs/source/overview.rst
Comment (caggle) [inline docs/source/overview.rst]: Maybe change this to "Implement schema validation on incoming events".

File: docs/source/overview.rst
Comment (caggle) [inline docs/source/overview.rst:39]: I noticed something in the single source diagram. On step 3 of the manual workflow, we say: "Create a ZIP archive of these config files" whereas this is actually not required for the single source code repo usage. It is only required for the multi-source scenario if code and config are to be kept in separate repositories (and in doing so CloudFormation basically creates a skeleton CodeCommit repo with template files under `config` directory).

File: docs/source/overview.rst
Comment (caggle) [inline docs/source/overview.rst:49]: As a continuation of the previous comment...

In this diagram, on the right hand side (cloudformation process flow), between 2nd and 3rd step, we should add "Creates a codecommit repo from the config archive stored in your helper bucket", because that's currently what the multi-source template does.

File: docs/source/usage/preparation.rst
Comment (caggle) [inline docs/source/usage/preparation.rst:1]: I really liked this section! I think it glues concepts together really well.

File: docs/source/usage/examples.rst
Comment (caggle) [inline docs/source/usage/examples.rst:22]: Very minor nit: We can delete this `ssm:` line - both from docs and from the code - since we are not referring to any SSM parameter during build time (i.e. it's not needed in the framework's current state).

File: docs/source/usage/examples.rst
Comment (caggle) [inline docs/source/usage/examples.rst]: Consider changing to "Deploy Script Example".

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: "This guide uses Zoom as our service, and assumes you will be deploying the framework to a development/test environment for the first time. ..."

Is it better?

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: If we are using `dev` as stage in the beginning of the file, we should change all stage/environment variables to "dev"

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst]: Same as above.

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst:132]: I know I had written the bulk of this section originally, but this whole section assumes we are using the single source template which uses CodeCommit to house all code + config. And I did so because that's our proposed method to deploy our instance.

However, we should probably document all 3 deployment steps, as there are some important points we don't currently touch here. For instance: if a Github source is used as a single repo, we use a Github personal access token to poll for changes in Github to then trigger a build (and we are using Secrets Manager to store that token in AWS). Naturally none of that is mentioned here.

Another example here, the very first step below actually does not apply if we are using the multi-source template to deploy, as in that case the CodeCommit repo is created by the CloudFormation template during stack creation.

I can help documenting the other deployment steps, but need to actually perform them again to refresh my memory :)

**Review by caggle — CHANGES_REQUESTED** [review]
This is amazing stuff! It was a good read, and understandable.

I did point out some small and some larger issues however. I hope they make sense! And I am happy with helping out with some of them.

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst]: thanks! fixed in the next commit

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst]: done

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst:39]: Thanks for spotting this! I'll review and resolve

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst:49]: Thanks, will review and resolve!

Comment (Phrozyn) [reply docs/source/overview.rst:49]: I'm still on the fence on whether this is the best method for this, it feels clunky to put config data in an s3 bucket that is going into a codecommit repo, when it should only just live in the codecommit repo (this prevents possibly creating an s3 bucket with lax perms if secrets are ever placed in those configs for testing purposes or some other reason)
Lets revisit? But I'll go ahead and modify the diagram for now.

File: docs/source/usage/examples.rst
Comment (Phrozyn) [reply docs/source/usage/examples.rst]: done

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: I've adjusted this, let me know if what I've added suits you.

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: Done

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst]: Done

File: docs/source/usage/preparation.rst
Comment (Phrozyn) [reply docs/source/usage/preparation.rst:1]: Ty!

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst:49]: Diagram modified

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst:132]: I would definitely like some help in mapping out the differences, I think it would be beneficial to demonstrate what those are.

Comment (caggle) [reply docs/source/usage/deployment.rst:132]: I am going to give this a go for other deployment options.

File: docs/source/overview.rst
Comment (caggle) [reply docs/source/overview.rst:49]: AFAIK, there is no other way to create a codecommit repo with some files, other than putting the files in a zip archive in an S3 bucket first.

We could just have the CF template create repo without any code/files, or we could ditch the idea of creating the codecommit repo using CloudFormation completely (so it becomes a manual task as in the single source scenario).

If we choose the first option, the very first app deployment after the CI stack is deployed will fail (as there will be no config files). If we create it manually beforehand I think it will be OK (and is probably the best option)

File: docs/source/usage/deployment.rst
Comment (caggle) [reply docs/source/usage/deployment.rst:132]: In the last 2 commits, I gave this one a go. It's not complete yet, as I have not documented the deployment steps for multi-source scenario (with merge).

_Disclaimer: I suck at editing `.rst` files, I tried to use a preview in VS Code when making the edits. Apologies if it looks crappy, but with these commits I went for content instead of aesthetics._

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst:212]: I stopped documenting this before we decide whether or not we want to create the CodeCommit repo manually beforehand or let the CF template create it for us (as is). Once we decide that I am happy to provide steps here.

File: docs/source/usage/deployment.rst
Comment (caggle) [inline docs/source/usage/deployment.rst:135]: I changed the rest to toggle based on deployment options. It's just an idea, we don't have to do it this way. Maybe we should dedicate each deployment scenario their own .rst file and link to them in here instead.

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst:132]: No worries :) we can polish before merging! You're awesome!

Comment (Phrozyn) [reply docs/source/usage/deployment.rst:132]: So I noticed that you listed 'public_repo' as a permission for the github token, if the user has elected to use a private repo, then this would not work, correct?

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst:135]: I love the toggle <3

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [inline docs/source/usage/deployment.rst]: For this line, as the access token can only be one per repo, this should likely be tied to the repo and the stack using it, otherwise with multiple deployments it will be unweildy

File: docs/source/usage/deployment.rst
Comment (caggle) [reply docs/source/usage/deployment.rst:132]: I don't think so. But I think I missed adding a scope, the parent `repo` scope, for which the description in Github says: "Full control of private repositories", as below. I will add this.

image

File: docs/source/usage/deployment.rst
Comment (caggle) [reply docs/source/usage/deployment.rst]: I don't think the personal access token is per repo - It's an account setting, so you can have a single personal access token with required scopes to access multiple repos.

Having said this, you are probably right in saying, we would like to have a personal access token per each stack, because otherwise if a token is leaked/rotated, every other stack sourcing from Github will be affected.

I will make a note of this in the documentation, and also change the relevant CF templates.

File: docs/source/usage/deployment.rst
Comment (caggle) [reply docs/source/usage/deployment.rst:132]: I added a pic instead to the relevant section in the next commit.

File: docs/source/usage/deployment.rst
Comment (caggle) [reply docs/source/usage/deployment.rst]: The next few commits address this, please review @Phrozyn.

Next up I will attempt to document the deployment using multi source template.

Comment (caggle) [reply docs/source/usage/deployment.rst]: I gave it a shot, while fixing some other inconsistencies in other deployment methods. I hope it is good.

**Comment by Phrozyn** [comment]
The template changes should not go into this documentation commit, those should go in a separate PR. I'm trying to keep documentation separate from code.
There are two other PR's in relation to the code base that can be used, or a new PR opened to address the changes you've committed here.
Mainly because the permissions PR has additional changes that this one is going to revert if we push it. So having it here will create conflicts.

**Comment by caggle** [comment]
Yeah, makes a lot of sense. Will revert those commits and then apply them to #23 instead.

File: docs/source/overview.rst
Comment (caggle) [inline docs/source/overview.rst:49]: Diagram does not load here :( When the link is clicked it also returns 404.

File: docs/source/usage/preparation.rst
Comment (caggle) [inline docs/source/usage/preparation.rst:21]: Good section, concise and understandable.

File: docs/source/usage/preparation.rst
Comment (caggle) [inline docs/source/usage/preparation.rst]: The word "that" here before "must" is probably not necessary.

File: docs/source/usage/preparation.rst
Comment (caggle) [inline docs/source/usage/preparation.rst:50]: Maybe here we should say: "GitHub Framework Public Repo" to ensure distinction from the custom Github repo above.

File: docs/source/usage/preparation.rst
Comment (caggle) [inline docs/source/usage/preparation.rst:57]: With this https://github.com/mozilla/mozdef-event-framework/pull/23/commits/5173cf49fb2046d22237a2388cc714024cc1696d in #23 we changed it so that a codecommit repo is not created by the multi source template anymore (needs to be created manually). We should delete the `note` section below.

**Review by caggle — COMMENTED** [review]
Ready to merge after addressing pending comments. Thanks!

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst:49]: Not exactly sure why this change would cause it not to load, am trying to troubleshoot this now.

File: docs/source/usage/preparation.rst
Comment (Phrozyn) [reply docs/source/usage/preparation.rst]: modified to read: The "env:" variables must be populated

File: docs/source/usage/preparation.rst
Comment (Phrozyn) [reply docs/source/usage/preparation.rst:57]: removed

File: docs/source/usage/preparation.rst
Comment (Phrozyn) [reply docs/source/usage/preparation.rst:50]: Changed to say A CodeCommit or separate Github repository

File: docs/source/overview.rst
Comment (caggle) [reply docs/source/overview.rst:49]: The image now loads, but clicking on them for full size now returns 404 because they link to: https://mozdef-event-framework.readthedocs.io/en/rtd_buildout/overview.rst

File: docs/source/usage/deployment.rst
Comment (Phrozyn) [reply docs/source/usage/deployment.rst:212]: I believe this was resolved by creating the repo first, instead of the CF doing it.

File: docs/source/overview.rst
Comment (Phrozyn) [reply docs/source/overview.rst:49]: I had added a target: but it's removed now.

**Review by caggle — APPROVED** [review]

**Comment by caggle** [comment]
Amazing work for documenting the framework, thank you!

**Pull request merged**
